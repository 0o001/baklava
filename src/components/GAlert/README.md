# GAlert

## Slots

| Name      | Description |
| --------- | ----------- |
| `default` | &nbsp;      |

## Props

| Name        | Type                     | Description | Default     |
| ----------- | ------------------------ | ----------- | ----------- |
| `variant`   | `String`                 |             | `"error"`   |
| `icon`      | `String`                 |             | `undefined` |
| `icon-size` | `Number` &#124; `String` |             | `"16px"`    |
| `show-icon` | `Boolean`                |             | `true`      |
| `border`    | `Boolean`                |             | `false`     |

## Computed Properties

| Name       | Type      | Description                           |
| ---------- | --------- | ------------------------------------- |
| `classes`  | `object`  | **Dependencies:** `variant`, `border` |
| `iconType` | `boolean` | **Dependencies:** `icon`, `variant`   |

