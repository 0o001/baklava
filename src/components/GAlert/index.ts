import GAlert from './GAlert.vue';

export default GAlert;
