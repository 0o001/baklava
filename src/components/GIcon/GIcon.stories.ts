import { storiesOf } from '@storybook/vue';
import GIcon from './GIcon.vue';

storiesOf('GIcon', module)
  .addParameters({ component: GIcon })
  .add('Example', () => ({
    components: { GIcon },
    template: `<div> <GIcon name="alert-circle" strokeWidth='2px' /> <GIcon stroke="black" size="44px" fill="var(--orange-100)" name="star" /> </div>`,
  }))