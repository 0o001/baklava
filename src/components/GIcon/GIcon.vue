<template>
  <svg
    v-bind="this.attrs"
    v-on="$listeners"
    :class="classNames"
    :dataName="this.name"
    :dataTags="tags"
    :dataType="this.name"
    v-html="icon.contents"
  />
</template>

<script lang="ts">

import feather from 'feather-icons';

export default {
  name: 'GIcon',
  props: {
    animation: {
      type: String,
      default: undefined,
    },
    animationSpeed: {
      type: String,
      default: undefined,
    },
    fill: {
      type: String,
      default: 'none',
    },
    size: {
      type: [Number, String],
      default: '24px',
    },
    stroke: {
      type: String,
      default: 'currentColor',
    },
    strokeLinecap: {
      type: String,
      default: 'round',
    },
    strokeLinejoin: {
      type: String,
      default: 'round',
    },
    strokeWidth: {
      type: [Number, String],
      default: 2,
    },
    tag: {
      type: String,
      default: 'i',
    },
    name: {
      type: String,
      required: true,
      validator (name) {
        if (!feather.icons[name]) {
          throw new Error(`"${name}" is not an available icon type.`);
        }
        return true;
      },
    },
    color: {
      type: String,
      default: undefined,
    },
  },
  computed: {
    icon () {
      return feather.icons[this.name];
    },
    tags () {
      return feather.icons[this.name] ? feather.icons[this.name].tags : '';
    },
    classNames () {
      const classes = {
        'g-icon': true,
        feather: true,
        [`feather--${this.name}`]: true,
        [`feather--${this.animation}`]: this.animation,
        [`feather--${this.animationSpeed}`]: this.animationSpeed,
      };
      return classes;
    },
    attrs () {
      return {
        ...this.icon.attrs,
        fill: this.fill,
        height: this.size,
        stroke: this.color ? `var(--${this.color})` : this.stroke,
        'stroke-linecap': this.strokeLinecap,
        'stroke-linejoin': this.strokeLinejoin,
        'stroke-width': this.strokeWidth,
        width: this.size,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@keyframes feather--spin {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
.feather {
  display: inline-block;
  overflow: hidden;
  &--spin {
    animation: feather--spin 2s linear infinite;
  }
  &--pulse {
    animation: feather--spin 2s infinite steps(8);
  }
  &--slow {
    animation-duration: 3s;
  }
  &--fast {
    animation-duration: 1s;
  }
  &__content {
    display: block;
    height: inherit;
    width: inherit;
  }
}
</style>
