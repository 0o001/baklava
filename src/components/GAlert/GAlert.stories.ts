import { storiesOf } from '@storybook/vue';
import GAlert from './GAlert.vue';
import GText from '../GText/GText.vue';
import GBox from '../GBox/GBox.vue';
import GIcon from '../GIcon/GIcon.vue';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

const stories = storiesOf('GAlert', module);
stories.addDecorator(withKnobs);
stories
  .addParameters({ component: GAlert })
  .add('Default', () => ({
    props: {
      text: {
        default: text('Text', '2'),
      },
      border: {
        default: boolean('Border', false),
      },
    },
    components: { GAlert, GIcon, GText, GBox },
    template: `
      <div style='padding:30px; display: flex; justify-content: space-between; flex-direction: column; height: 400px'>
        <GAlert variant="error" :border="border">
          Error
        </GAlert>
        <GAlert variant="warning" :border="border">
          Warning
        </GAlert>
        <GAlert variant="success" :border="border">
          Success
        </GAlert>
        <GAlert variant="info" :border="border">
          Info
        </GAlert>
        <GAlert variant="info" icon-size="24px">
          <GBox> Line 1 </GBox>
          <GBox> Line 2 </GBox>
          <GBox> Line 3 </GBox>
        </GAlert>
        <GAlert variant="info" :show-icon="false">
          <GBox> Without Icon </GBox>
        </GAlert>
      </div>`,
  }));
