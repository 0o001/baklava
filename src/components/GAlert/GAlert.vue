<template>
  <GBox
    :class="classes"
    flex
    align="items-center"
  >
    <GIcon
      v-if="showIcon"
      :name="iconType"
      :size="iconSize"
    />
    <GBox class="content">
      <slot />
    </GBox>
  </GBox>
</template>

<script>

import GIcon from '../GIcon/GIcon.vue';
import GBox from '../GBox/GBox.vue';
import $_ from '../../utils/dashUtil.ts';

export default {
  name: 'GAlert',
  components: { GBox, GIcon },
  props: {
    variant: {
      type: String,
      default: 'error',
    },
    icon: {
      type: String,
      default: undefined,
    },
    iconSize: {
      type: [Number, String],
      default: '16px',
    },
    showIcon: {
      type: Boolean,
      default: true,
    },
    border: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classes () {
      return {
        'g-alert': true,
        [$_(this.variant)]: true,
        '-border': this.border,
      };
    },

    iconType () {
      const icons = {
        alert: 'alert-circle',
        warning: 'alert-triangle',
        success: 'check-circle',
        info: 'info',
      };

      return this.icon || icons[this.variant] || 'alert-circle';
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-alert {
    width: 100%;
    padding: 15px;
    border-radius: 4px;
    font-size: var(--font-size-12);

    .content {
      font-weight: 500;
    }

    .g-icon {
      margin-right: 10px;
    }

    &.-success {
      background-color: var(--green-900);
      color: var(--green-500);

      &.-border {
        border: 1px solid var(--green-300)
      }
    }

    &.-info {
      background-color: var(--blue-900);
      color: var(--blue-500);

      &.-border {
        border: 1px solid var(--blue-300)
      }
    }

    &.-warning {
      background-color: var(--orange-900);
      color: var(--orange-500);

      &.-border {
        border: 1px solid var(--orange-300)
      }
    }

    &.-error {
      background-color: var(--red-900);
      color: var(--red-500);

      &.-border {
        border: 1px solid var(--red-300)
      }
    }
  }
</style>
