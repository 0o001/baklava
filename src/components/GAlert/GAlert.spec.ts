import { shallowMount, ThisTypedShallowMountOptions, Wrapper } from '@vue/test-utils';
import GAlert from './';
import GIcon from '../GIcon/GIcon.vue';

describe('GAlert', () => {
  let wrapper: Wrapper<GAlert>;
  const renderWrapper = (options: ThisTypedShallowMountOptions<GAlert> = {}) => {
    wrapper = shallowMount(GAlert, options);
  };

  it('should match snapshot with default style', () => {
    renderWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not render GIcon if showIcon is false', () => {
    renderWrapper({
      propsData: {
        showIcon: false,
      },
    });

    expect(wrapper.findComponent(GIcon).exists()).toBeFalsy();
  });

  it('should has prop icon if icon is given as prop', () => {
    renderWrapper({
      propsData: {
        icon: 'user',
      },
    });

    const iconComponent = wrapper.findComponent(GIcon);

    expect(iconComponent.exists()).toBeTruthy();
    expect(iconComponent.attributes().name).toBe('user');
  });

  it('should has variant icon type if given variant is valid', () => {
    renderWrapper({
      propsData: {
        variant: 'success',
      },
    });

    const iconComponent = wrapper.findComponent(GIcon);

    expect(iconComponent.exists()).toBeTruthy();
    expect(iconComponent.attributes().name).toBe('check-circle');
  });

  it('should has alert-circle icon if not variant icon type and given variant is invalid', () => {
    renderWrapper({
      propsData: {
        variant: 'invalid',
      },
    });

    const iconComponent = wrapper.findComponent(GIcon);

    expect(iconComponent.exists()).toBeTruthy();
    expect(iconComponent.attributes().name).toBe('alert-circle');
  });

  it('should has -alert class if border prop is given', () => {
    renderWrapper({
      propsData: {
        variant: 'success',
        border: true,
      },
    });

    expect(wrapper.classes('-border')).toBeTruthy();
  });
});
