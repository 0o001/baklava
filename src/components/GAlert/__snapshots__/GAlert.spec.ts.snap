// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GAlert should match snapshot with default style 1`] = `
<gbox-stub
  align="items-center"
  class="g-alert -error"
  flex="true"
>
  <gicon-stub
    fill="none"
    name="alert-circle"
    size="16px"
    stroke="currentColor"
    strokelinecap="round"
    strokelinejoin="round"
    strokewidth="2"
    tag="i"
  />
   
  <gbox-stub
    class="content"
  />
</gbox-stub>
`;
